llama_index/agent/openai_agent.py

```diff
@@ -145,10 +244,64 @@ async def achat(
             response = await self._llm.achat(chat_history, functions=functions)
             ai_message = response.message
             chat_history.append(ai_message)
-            function_call = ai_message.additional_kwargs.get("function_call", None)
+            function_call = self._get_latest_function_call(chat_history)
 
         return Response(ai_message.content)
 
+    def astream_chat(
+        self, message: str, chat_history: Optional[List[ChatMessage]] = None
+    ) -> AsyncGenerator[StreamingChatResponse, None]:
+        chat_history = chat_history or self._chat_history
+        tools, functions = self._init_chat(chat_history, message)
+
+        async def _astream_chat(
+            chat_history: List[ChatMessage],
+        ) -> AsyncGenerator[StreamingChatResponse, None]:
+            # TODO: Support forced function call
+            chat_stream_response = StreamingChatResponse(
+                await self._llm.astream_chat(chat_history, functions=functions)
+            )
+
+            # Get the response in a separate thread so we can yield the response
+            thread = Thread(
+                target=chat_stream_response.write_response_to_history,
+                args=(chat_history,),
+            )
+            thread.start()
```

Comment:
What do you think about just about having concurrency here by submitting this as a task on the asyncio event loop, rather than using threads? e.g. [`asyncio.create_task`](https://docs.python.org/3/library/asyncio-task.html#creating-tasks). Would that be possible?

Two reasons I can think of for opting for asyncio rather than threads:
- Since I'm not aware of any other part of the library spinning up new threads, it may come as a surprise to users to find threads being spun up in their processes.
- Since this may well be deployed as part of a chat server application that handles multiple chats at once, there may be utilization limits to how many threads can be spun up at once.

Reply:
We use threads for other streaming capabilities actually

https://github.com/jerryjliu/llama_index/blob/fca0da48158ab0f9cb07cfe335ca3ea3892d4c40/llama_index/llms/langchain.py#L62

https://github.com/jerryjliu/llama_index/blob/fca0da48158ab0f9cb07cfe335ca3ea3892d4c40/llama_index/llms/huggingface.py#L161

I can try swapping for asyncio, but not 100% sure it will end up working properly. I can let you know

Reply:
@sourabhdesai yea, asyncio doesn't seem to be able to let me both stream into the queue and allow the user to read the queue at the same time 🤔 

Tried putting the asyncio call both before and after the yield (also, using asyncio for non-async function calls breaks notebooks without using nest async, which is a little confusing 🤔 )
```
yield chat_stream_response
asyncio.run(chat_stream_response.write_response_to_history(chat_history))
```

Reply:
AFAIK it won't work with just another asyncio loop. I tried a bit with when I was trying to do it for`llama_index/llms/langchain.py`.